tests/test_recipe.py

```diff
@@ -657,6 +657,75 @@ def test_recipe_as_subquery(self):
         self.assertRecipeCSV(r2, "pop\n3147355.0")
 
 
+class TestRecipeIngredientsFromMultipleTables(RecipeTestCase):
+    def test_same_table(self):
+        from sqlalchemy.sql.expression import join
+
+        # Shelves with multiple tables don't raise
+        # exceptions when you select ingredients that all come
+        # from the same table.
+        self.shelf = self.mixed_shelf
+
+        recipe = self.recipe().metrics("score").dimensions("username")
+        self.assertRecipeSQL(
+            recipe,
+            """SELECT scores.username AS username,
+       avg(scores.score) AS score
+FROM scores
+GROUP BY username""",
+        )
+        self.assertRecipeCSV(
+            recipe,
+            """
+            username,score,username_id
+            annika,85.0,annika
+            chip,90.0,chip
+            chris,80.0,chris
+            """,
+        )
+
+    def test_multi_table(self):
+        self.shelf = self.mixed_shelf
+
+        with self.assertRaises(BadRecipe):
+            recipe = self.recipe().metrics("score").dimensions("tagscorestestid")
+            # Exception is raised only on query generation
+            recipe.to_sql()
+
+        # We can optionally allow multiple tables to be joined
+        recipe = (
+            self.recipe()
+            .metrics("score")
+            .dimensions("tagscorestestid")
+            .allow_multiple_tables(True)
+        )
+        # In this case, the dimension contains filtering criteria that allow it
+        # to jion tagscores and scores
+        self.assertRecipeSQL(
+            recipe,
+            """
+            SELECT tagscores.testid AS tagscorestestid,
+                avg(scores.score) AS score
+            FROM tagscores,
+                scores
+            WHERE tagscores.username = scores.username
+            GROUP BY tagscorestestid
+            """,
+        )
+        self.assertRecipeCSV(
+            recipe,
+            """
+            tagscorestestid,score,tagscorestestid_id
+            1,80.0,1
+            2,90.0,2
+            3,90.0,3
+            4,90.0,4
+            5,85.0,5
+            6,85.0,6
+            """,
+        )
```

Comment:
can you add a test here that selects  multiple columns from `tagscores`? Maybe these cases:

- selecting two different dimensions from tagscores in addition to "score"
- selecting a dimension and a column from tagscores in addition to "score" (I see you defined `tagscorestest_cnt` but haven't used it in a test yet)

Reply:
@radix Sounds good.